proposals/kroxylicious-operator-api-alternative.md

```diff
@@ -0,0 +1,654 @@
+    resource:
+       bootstrap: upstream:9092
+       protocol: TLS
+       nodeIdRanges:
+      - name: mybrokers
+        range:
+          startInclusive: 0
+          endExclusive: 3
+
+  filterRef:
+  - group: filter.kroxylicious.io
+    kind: Filter
+    name: encryption 
+```
+
+What would operator create:
+* kroxylicious deployment, 1 replica
+* proxy relies on port based routing  (needs new PortPerNode scheme that allows the Operator to control the ports precisely.  This is needed to prevent the potential for "crossed-lines" during reconcilations.)
+* 1 ClusterIP services (with n+1 ports) - operator needs to _deterministically_ assign a block of ports for this VC and create the port/target mapping in the Service accordingly.
+* Kafka Clients connect to serviceaddress:9082 
+
+What the user would provide:
+* secret containing server certificate with the services
+
+### On Cluster Traffic - tls downstream & upstream - varation using OpenShift Cluster CA generated cert
+
+Proxy and ProxyIngress as above
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: VirtualCluster
+metadata:
+  name: mycluster
+spec:
+  proxyReference:
+     ...
+
+  ingress:
+  - name: myclusterip
+   infrastructure:
+       annotations:
+         service.beta.openshift.io/serving-cert-secret-name=myservingcert   # Tells openshift to generate a secret
+    tls:
+      certificateRef:
+         name: myservingcert
+ 
+  targetCluster:
+     ...
+
+  filterRef:
+     ...
+```
+
+What would operator create:
+(as above)
+
+What the user would provide:
+(nothing - openshift is generating the serving cert for us and providing DNS)
+
+## Off Cluster Traffic (OpenShift Route)
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: Proxy
+metadata:
+  name: myproxy
+spec: {}
+```
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: ProxyIngress
+metadata:
+  name: myopenshiftroute
+spec:
+  proxyRef:
+    name: myproxy
+
+  openShiftRoute: {}
+```
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: VirtualCluster
+metadata:
+  name: mycluster
+spec:
+  proxyReference:
+    name: myproxy
+    kind: Proxy
+    group: proxy.kroxylicious.io
+
+  ingress:
+  - name: myopenshiftroute
+    tls:
+      certificateRef:
+         ...
+ 
+  targetClusterRef:
+     resource:
+       bootstrap: upstream:9092
+       protocol: TLS
+       nodeIdRanges:
+      - name: mybrokers
+        range:
+          startInclusive: 0
+          endExclusive: 3
+
+  filterRef:
+  - group: filter.kroxylicious.io
+    kind: Filter
+    name: encryption 
+```
+
+What would operator create:
+* kroxylicious deployment, 1 replica
+* proxy relies on SNI based routing (use SNI scheme)
+* ClusterIP service called `myopenshiftroute-service` (shared)
+* n+1 OpenShift Routes configured for pass-through. The `to` points to the `myopenshiftroute-service`.
+
+Note: that the operator would write the virtualcluster proxy config based on the hostnames assigned to the Routes by the Ingress Controller.
```

Comment:
We need the openshift routes to match the virtual cluster broker address pattern. Would it be the Operator's responsibility to set up the Route hosts, or should the users configure something like `broker-${nodeId}.abc.def.${ingressDomain}`?

Reply:
> We need the openshift routes to match the virtual cluster broker address pattern.

I don't think this is something we'd surface to users.

I was imagining it would be the Operator's responsibility. It would create routes with names in a well-known form:

broker-${nodeId}-${virtualcluster}
bootstrap-${virtualcluster}

and the Ingress Operator would assign the hostnames.

broker-${nodeId}-${virtualcluster}-${namespace}.${ingressDomain}
bootstrap-${virtualcluster}-${namespace}.${ingressDomain}

We'd need to document how the name will be formed so that the user can generate their certificate.

There's a potential to support vanity names (by setting the Routes spec.host field), but I think we'd leave that for an RFE.